src/AsyncBridge/AsyncCompatLibExtensions.cs

```diff
@@ -85,5 +97,231 @@ public static ConfiguredTaskAwaitable ConfigureAwait(this Task task, bool contin
+    /// cref="T:System.Threading.CancellationTokenSource"/> has been disposed.
+    /// </exception>
+    /// <exception cref="T:System.ArgumentOutOfRangeException">
+    /// The exception thrown when <paramref name="delay"/> is less than -1 or 
+    /// greater than Int32.MaxValue.
+    /// </exception>
+    /// <remarks>
+    /// <para>
+    /// The countdown for the delay starts during this call.  When the delay expires, 
+    /// this <see cref="T:System.Threading.CancellationTokenSource"/> is canceled, if it has
+    /// not been canceled already.
+    /// </para>
+    /// <para>
+    /// Subsequent calls to CancelAfter will reset the delay for this  
+    /// <see cref="T:System.Threading.CancellationTokenSource"/>, if it has not been
+    /// canceled already.
+    /// </para>
+    /// </remarks>
+    public static void CancelAfter(this CancellationTokenSource cancelSource, TimeSpan delay)
+    {
+        if (cancelSource == null)
+            throw new ArgumentNullException(nameof(cancelSource));
+
+        if (delay < _timeoutInfinite)
+            throw new ArgumentOutOfRangeException(nameof(delay));
+
+        if (cancelSource.IsCancellationRequested)
+            return;
+
+        Timer MyTimer = null;
+
+        // CTS claims it's thread-safe for all methods, so we should be the same.
+        // If someone calls CancelAfter concurrently for the same CTS, we won't explode or cause a memory leak
+        do
+        {
+            if (_cancelTimers.TryGetTimer(cancelSource, out var CurrentTimer))
+            {
+                if (MyTimer != null)
+                    MyTimer.Dispose();
+
+                MyTimer = CurrentTimer;
+                break; // Reset the existing timer to the new delay
+            }
+
+            // An active timer doesn't hold a strong reference to itself, so adding the CTS as state won't hold it alive
+            MyTimer = new Timer(OnCancelAfterTimer, cancelSource, Timeout.Infinite, Timeout.Infinite);
+
+        } while (!_cancelTimers.TryAddTimer(cancelSource, MyTimer));
+
+        try
+        {
+            MyTimer.Change(delay, _timeoutInfinite);
+        }
+        catch (ObjectDisposedException)
+        {
+        }
+    }
+
+    private static void OnCancelAfterTimer(object state)
+    {
+        var CancelSource = (CancellationTokenSource)state;
+
+        if (!_cancelTimers.TryRemoveTimer(CancelSource, out var MyTimer))
+            return;
+
+        MyTimer.Dispose();
+
+        try
+        {
+            CancelSource.Cancel();
+        }
+        catch (ObjectDisposedException) // If the cancellation token has been disposed of, ignore the exception
+        {
+        }
+    }
+
+#if NET35
+    private struct WeakCTS : IEquatable<WeakCTS>
+    {
+        private readonly int _hashCode;
+
+        internal WeakCTS(CancellationTokenSource cts)
+        {
+            _hashCode = cts.GetHashCode();
+            CTS = new WeakReference(cts);
+        }
```

Comment:
If this struct ever became public, wouldn't it make sense for the constructor and the `CTS` property to be public too?

Reply:
I can't see why it'd ever become public - it exists entirely to support a weak reference to a CancellationTokenSource inside a dictionary, in order to not hold the CTS alive if it's disposed of.

Reply:
My point is, why mark some members internal if you have no plan to make the type public? I found it distracting.

Reply:
> I use internal because if you ever split the nested class out into its own file as an internal (so it's no longer nested) you'd mark those methods as internal too.

Except if I did ever split it out, I wouldn't do that. 😄 My preference is that it's more DRY to specify the visibility of the whole class in one place. Marking `internal` twice would suggest to me that there's a separate *reason* to access those methods than to access the whole class.
If you take this one step further and make it public, not unrealistic in general, the class is useless without those methods. Even though we aren't going to make it public, the thought experiment reveals the real complexity or simplicity of the situation (two logical access scopes or one).

It's doing no harm so it's up to you. It's fun to discuss though!